CIP-0158? | Cardano URIs - Browse Authority

Crypto2099 opened the pull request:
This CIP seeks to define the `browse` authority for `web+cardano` URIs. The intent here being to easily enable deep-linking capabilities for Cardano mobile wallets to launch their native embedded browser to the specified application or URL. This should be particularly useful in things like point-of-sale systems and for sharing applications in the wild versus the current process which can be rather clunky requiring users to:
1. Open the wallet
2. Find the wallet's application browser
3. Search or manually type a URL to the application

*Note:* Due to personal time constraints I did use some AI assistance in drafting some of the language of this pull request but the concept behind the scheme is mine (and inspired by the existing work of @francisluz and Begin Wallet) and all language has been reviewed and edited by myself personally.

[View Rendered Document in Branch](https://github.com/Crypto2099/CIPs/tree/cardano-uris-browse-authority/CIP-0158/README.md)

Crypto2099 committed:
First draft CIP for creating a `browse` URI scheme to extend web+cardano and deep-link capabilities for application support.

rphair commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,163 @@
+---
+CIP: XXXX
+Title: Cardano URIs | Browse Application
```
```suggestion
Title: Cardano URIs - Browse Application
```
CIP titles appear in Markdown tables in several contexts, so `|` disrupts the formatting — and escaping that character isn't recognised downstream from the YAML context — so we've decided to use `-` instead for the delimiter here as in several other categorised CIPs (also changed in PR title).

rphair commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,163 @@
+* Direct the wallet to launch its embedded browser or app view
+* Navigate directly to the intended dApp page or state without manual user input
+
+This reduces friction, improves dApp discoverability, and creates a smoother
+experience for both developers and end-users, particularly in mobile and
+cross-device contexts.
```
```suggestion
This reduces friction, improves dApp discoverability, and creates a smoother
experience for both developers and end-users, particularly in mobile and
cross-device contexts.
```
After this paragraph I would also be happy to see another motivation (a Security benefit, given that the potential Security liabilities are also addressed):

This scheme can also avoid naive user search for dApps, wallets, token wrappers, and many other apps whose Internet search terms, Telegram groups, etc. have been targeted by malicious fakes.

Offering them a canonical URL mighn't eliminate the user responsibility to keep track of correct URLs completely (since they would still need to be sure they could trust where these protocol links were coming from) but being able to coordinate distribution safely (through the reverse domain name) I believe would prevent the common catastrophic scenario of the user misdirected to a malicious app & losing their funds there.

rphair replied:
thanks @Crypto2099 - this is fully resolved by https://github.com/cardano-foundation/CIPs/pull/1058/commits/7bf2e221ab8c9cd846be6d4a85a7ac6be8358a43 & includes the points that we talked about at the meeting (mentioned again in https://github.com/cardano-foundation/CIPs/pull/1058#pullrequestreview-3048266071).

rphair reviewed (commented):
thanks @Crypto2099 - I like too many things about this proposal to enumerate them all here 😅 so I hope it will suffice to say I will get behind it 100%, though also with care to ensure through peer review that:
* the URL scheme is sufficient to cover any needed parameter passing into targeted applications with no ambiguity;
* any presented scenarios for insecurity, obfuscation of malicious apps, or other abuse are explored & documented (even if they seem trivial at the time).

Looking forward to hearing your introduction of the CIP from the agenda of our next meeting 🚀 https://hackmd.io/@cip-editors/116

Crypto2099 committed:
Update title

Co-authored-by: Robert Phair <rphair@cosd.com>

Ryun1 commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,163 @@
+experience for both developers and end-users, particularly in mobile and
+cross-device contexts.
+
+## Specification
```
note from meeting: worth adding the versioning to this standard now

rphair commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,163 @@
+---
+CIP: XXXX
```
```suggestion
CIP: 158
```
🎉

rphair reviewed (commented):
Confirmed at yesterday's CIP meeting that this satisfies a clear & present need in the community — especially for moving quickly from presentation to wallet environments, and from desktop to mobile environments — also confirming the additional security motivation I mentioned in https://github.com/cardano-foundation/CIPs/pull/1058#discussion_r2217097994.

Personally I expect to ✅ this after a mental walk-through of some deployment scenarios — somewhat naively since @Crypto2099 I don't have your experience in the targeted environments like conventions and token drops — and hope to do this by the end of this week: especially cross referencing with CIP-0013 family members & https://github.com/cardano-foundation/CIPs/pull/843.

MadOrkestra commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,163 @@
+* **Global uniqueness:** avoids collisions.
+* **Decentralized management:** no central registry needed.
+* **Developer familiarity:** matches Android, Apple, Java ecosystems.
+* **Verifiability:** can tie domains to apps via `.well-known` or DNS records.
```
Just a random thought on this: is there an opportunity here to propose a mechanism for enhanced security / automatic whitelisting across multiple wallets? As it stands now, almost all wallets do some kind of individual white-listing for domains or have their own pre-filled catalogue of Cardano dApps, which makes sense to avoid scams, but also is very centralized as it is done on a per wallet basis.

It might be worth thinking about this for a second as this CIP will - while greatly improving the UX - basically opens up wallets for all kinds of shenanigans and I can already see the reluctance to implement this.

Ofc this can't be the next centralized registry, but what about an on-chain majority vote multisig for wallets/trusted entities where domains can get verified and whitelisted on chain?

rvcas commented:
yes please

Crypto2099 commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,163 @@
+* **Scheme (REQUIRED):** `https`, `http`, `ipfs`, etc.
+* **Namespaced domain (REQUIRED):** reverse domain name (e.g., `fi.sundae.app`)
+* **App-specific path (OPTIONAL):** full path within the app
+* **Query parameters (OPTIONAL):** optional, passed as-is to the app
```
Would this be specifically for URL Query parameters as the path could obviously not contain spaces and still remain valid?

MadOrkestra replied:
I mean, its certainly not impossible to create paths with unencoded special-characters, but you'd have to be nuts.

Crypto2099 replied:
I have added a note here just explicitly pointing out that both the app-specific path and query parameters should both be URL encoded. Better safe than sorry

MadOrkestra replied:
good to go!

Crypto2099 commented:
> Otherwise: great job as always by @Crypto2099 - I am working on an Affiliate Platform rn, which will make use of this as soon as we have the first mobile wallet implementing this :)
> 
> One thought though: Maybe we go with //launch instead of //browse? This would future proof this for further applications, because I can see use-cases where this could be used to _launch_ other wallet features than the in-app browser without the need of a new CIP?

If we decided to change this, how would we indicate that, in our case, we want to launch the built in dApp browser instead of... idk... a built-in "Purchase ADA" function?

Crypto2099 commented:
> Just a random thought on the topic of verifying domains: is there an opportunity here to propose a mechanism for enhanced security / automatic whitelisting across multiple wallets? As it stands now, almost all wallets do some kind of individual white-listing for domains or have their own pre-filled catalogue of Cardano dApps, which makes sense to avoid scams, but also is very centralized as it is done on a per wallet basis.
> 
> It might be worth thinking about this for a second as this CIP will - while greatly improving the UX - basically open up wallets for all kinds of shenanigans and I can already see the reluctance to implement this.
> 
> Ofc this can't be the next centralized registry, but what about an on-chain majority vote multisig for wallets/trusted entities where domains can get verified and whitelisted on chain?

I have thought about this and I spoke, once upon a time quite some time ago when CIP-88 was first making the rounds, with a few folks at CF who were responsible for the off-chain token registry... This was specifically related to how we would confirm the veracity of smart contract minted tokens but the same concept stands. I wonder if there's something along the lines of CIP-0155 (#1033) in terms of special DNS entries and other security mechanisms that could be implemented to increase trust in a decentralized fashion w/o going full Black Mirror Social Credit dystopia...

Crypto2099 committed:
Update CIP Number

Co-authored-by: Robert Phair <rphair@cosd.com>

Crypto2099 committed:
Update pull request number in discussion header

Co-authored-by: Ryan <ryan.williams@intersectmbo.org>

MadOrkestra commented:
> > Otherwise: great job as always by @Crypto2099 - I am working on an Affiliate Platform rn, which will make use of this as soon as we have the first mobile wallet implementing this :)
> > One thought though: Maybe we go with //launch instead of //browse? This would future proof this for further applications, because I can see use-cases where this could be used to _launch_ other wallet features than the in-app browser without the need of a new CIP?
> 
> If we decided to change this, how would we indicate that, in our case, we want to launch the built in dApp browser instead of... idk... a built-in "Purchase ADA" function?

On second thought: lets forget about this. 

_(this is what crossed my mind: I could - as an App developer - kinda go web+cardano://launch/my-btc-protocol-identifier/purchase-bitcoin - but it makes no sense to put this into a _standard_ as multiple wallets would have to agree about this ofc)_

Crypto2099 committed:
Updates based on initial feedback
- Introduce versioning to URI scheme
- Clean up some grammatical errors
- Add explanation of security benefits and considerations section to the motivation
- Update some parts of implementation plan that have been completed.

rphair committed:
fix typo to clear review field

AlexDochioiu commented on CIP-XXXX/README.md:
````diff
@@ -0,0 +1,203 @@
+
+### URI Format
+
+```
+web+cardano://browse/<version>/<scheme>/<namespaced_domain>/<app-specific-path>?<parameters>
+```
+
+* **Authority (REQUIRED):** `browse`
+* **Version (REQUIRED):** `v1`
+* **Scheme (REQUIRED):** `https`, `http`, `ipfs`, etc.
+* **Namespaced domain (REQUIRED):** reverse domain name (e.g., `fi.sundae.app`)
+* **App-specific path (OPTIONAL):** full, url-encoded path within the app
+* **Query parameters (OPTIONAL):** optional, url-encoded query parameters passed
+  as-is to the app
````
Hey @MadOrkestra @Crypto2099 This is not ideal, and doesn't actually work on some android devices.

For the following: 
`web+cardano://browse/v1/https/store.jpg/listing/22978160`
<img width="3466" height="1052" alt="Image" src="https://github.com/user-attachments/assets/d09cb98f-f250-497c-8582-9ccffbdc48b8" />

Multiple android devices just do not like what's being parsed and remove the `/` after `http`:


<img width="1080" height="2280" alt="Image" src="https://github.com/user-attachments/assets/fed8636d-ca44-4fde-8424-7f4cccfb07a3" />

AlexDochioiu replied:
We do not plan to integrate it as such since it's not working reliably on all devices/platforms and is not technically the correct way to include URLs in a deep link. We'll instead implement the following:

`web+cardano://browse/v1?uri=https%3A%2F%2Fstore.jpg%2Flisting%2F22978160`

Where:
* **Scheme (REQUIRED):** `web+cardano`
* **Authority (REQUIRED):** `browse`
* **Version (REQUIRED):** `v1`
* **Query parameter `uri` (REQUIRED):** percent-encoded uri to open in the app

If URI query param has no scheme, `https` is assumed

MadOrkestra replied:
Hey Alex, good catch, I wasn't aware of this errr lets call it "feature" (I guess https://domain.com/https/something wouldn't work on a QR code either then?) - this needs to be mitigated for sure.

As for protocols I absolutely believe this CIP should support other protocols besides http and https. This is web3, seems very likely at some time we'll have pages running on IPFS or whatever other decentralized protocols come around in the next years.

rphair replied:
URI misinterpretation issue apparently resolved with @AlexDochioiu's suggestions (committed by editors) beginning here: https://github.com/cardano-foundation/CIPs/pull/1058/commits/6c6694a0ea625badfab38c40df038cb4fb13a438 - cc @MadOrkestra & please unresolve with more detail if this remains an issue with the updated CIP specification.

rphair reviewed (commented):
As promised / expected this is a tentatively positive review for this proposal after a more detailed read: in my understanding it doesn't lack anything in detail or proposed usage.

@Crypto2099 as also expected I believe the editors will have to wait until responses like that of @AlexDochioiu's https://github.com/cardano-foundation/CIPs/pull/1058#discussion_r2238970050 are accommodated: probably through a convergence process between in-progress implementations and the proposed ideal.  Will keep following those revisions and certainly ✅ and move this forward as soon as those details are settled somehow.

Crypto2099 committed:
Update CIP-XXXX/README.md

Update per Robert's editorial suggestion

Co-authored-by: Robert Phair <rphair@cosd.com>

rphair commented:
thanks @MadOrkestra - are you saying that `cardano-mobile-connect-demo` itself has a faulty CIP-0158 implementation?

@Ryun1 @perturbing if the dev community is proceeding to what they think are "CIP-0158" implementations (since there is no CIP-0158, as this is only a candidate, i.e. a draft) we should discuss no later than the next meeting whether this should just be merged as-is after some editor cleanup: so I've put it on Review there ().  The state of Cardano URI implementations is challenged enough without implementations being based on incomplete work & documents that never passed peer review (cc @Crypto2099).

rphair commented:
@Crypto2099 are you going to resume pressing this PR forward: preferably with invitations to peer review & consideration of current implementations?  In any case I think we need to confirm that this CIP has a maintainer, as well as advocates, while we are seeing these implementations emerge.  cc @Ryun1 @perturbing

rphair commented:
@MadOrkestra due to the stall this CIP has been in, I put it on the CIP meeting Review agenda today to discuss the possibilities of correcting it somehow (based on your https://github.com/cardano-foundation/CIPs/pull/1058#issuecomment-3852505095) enough for it to be realistically merged according to what is being done out in the real world... assuming any "browse" implementations meet best practice guidelines as other wallet devs might see it & might be interested in implementing themselves.

In this meeting we had already decided (https://github.com/cardano-foundation/CIPs/pull/1112#pullrequestreview-3817044094) to collect a list of wallet devs to tag over another standards matter.  I quickly outlined our CIP-0013 / CPS-0016 issues which are effectively stalled without wallet ecosystem participation, and suggested solving these are necessary to build a Cardano point-of-sale economy:
* https://github.com/cardano-foundation/CIPs/issues/836

as well as @Crypto2099's contribution in indexing all these protocols in CPS-0016.  Those present at the meeting believed we might solve some of these issues if you (@MadOrkestra) were willing to update this document & perhaps some others in support of the standardisation effort you've tried to promote with the issue above.

Is this something you can help with?  @Ryun1 and @perturbing have agreed to assemble a GitHub tag list of wallet representatives (even if not meeting the goal of re-convening the "Wallets Working Group") and so we might finally be able to consolidate this effort with a person like yourself as CIP co-author & continuing as advocate.

AlexDochioiu commented:
@rphair I was trying to avoid the suggestions thing, as it becomes pretty annoying both to do and to review 😅

AlexDochioiu commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,202 @@
+* Apply allowlist/blocklist and security policies:
+    * Warn on `http` and `localhost`.
+    * Allow `https` by default.
+    * Resolve `ipfs` via trusted gateway or native resolver.
```
```suggestion
* Suggested allowlist/blocklist and security policies:
    * Each wallet may optionally implement its own allowed list of trusted domains where navigation would not require explicit permissions
    * Each wallet may optionally implement a blacklist of known untrusted domains where the user would be shown a clear warning that the website is known to be malicious. With very explicit user permission user should still be allowed to navigate.
    * For domains outside of allowlist / blacklist, wallet may want to show a warning and request explicit permission before navigating. This is to prevent unwanted IP / device info / other data exfiltration which can be obtained simply by loading a webpage
```

AlexDochioiu commented on CIP-XXXX/README.md:
```diff
@@ -0,0 +1,202 @@
+    - Reference Implementation: https://github.com/crypto2099/cardano-uri-parser
+    - NPM Package: https://www.npmjs.com/package/cardano-uri-parser
+- [ ] Collaborate with at least one dApp team to generate example deep links and
+  test them live (e.g., at events or with public beta testers).
```
```suggestion
```
I don't think this is needed. This dApp provides value even without this (by sharing token pari DEX URIs / doing in-person events where we want users to open and register somewhere / etc)

P.s. I know those still count as dApp integration, I just don't think we need this as a pre-requisite for validating how meaningful this CIP is

rphair commented on CIP-XXXX/README.md:
````diff
@@ -0,0 +1,202 @@
+### URI Format
+
+```
+web+cardano://browse/<version>/<scheme>/<namespaced_domain>/<app-specific-path>?<parameters>
````
@MadOrkestra - right now the way @AlexDochioiu has corrected it above is consistent with the only other protocol using `/version` = CIP-0099 (see its [ABNF Grammar](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0099#abnf-grammar) & examples above it): i.e. no `/` after the version ID.

Other protocols in [CPS-0016 > Registered URI Authorities](https://github.com/cardano-foundation/CIPs/blob/master/CPS-0016/README.md#registered-uri-authorities)) `//transaction` and `//block` (covered by the same CIP) say they support versioning: but versioning _is not_ in that CIP's URI syntax (see [CIP-0107 > Specification](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0107#specification)).  So as of now, only `//claim` has a `/version` field: and it has no `/` after the version ID.

At the time CPS-0016 was written, I think they were thinking of adding versioning to the URI strings for `//transaction` and `//block` but probably ruled it out because flexibility in the syntax for these immutable objects was considered not to be necessary.

So technically the statements "supports versioning (?)" should be removed from those CIP-0107 protocols in CPS-0016... unless it's supported out in the real world: in which case CIP-0107 itself should be updated by putting versioning into its URI syntax 😅 ... please post here if you think that's the case.

In any case, ~once all are in agreement I'm ready to commit this change~ I'll resolve & commit this to also assure consistency with the further corrections below which are about to be committed (before further review).

rphair committed:
update URI format

Co-authored-by: Alexandru Dochioiu <38853913+AlexDochioiu@users.noreply.github.com>

rphair committed:
update syntax parameters, including complete target URI encoding

Co-authored-by: Alexandru Dochioiu <38853913+AlexDochioiu@users.noreply.github.com>

rphair committed:
add Alex Dochioiu @ VESPR as co-author

Co-authored-by: Alexandru Dochioiu <38853913+AlexDochioiu@users.noreply.github.com>

rphair committed:
update example web app

Co-authored-by: Alexandru Dochioiu <38853913+AlexDochioiu@users.noreply.github.com>

rphair committed:
update example IPFS app

Co-authored-by: Alexandru Dochioiu <38853913+AlexDochioiu@users.noreply.github.com>

rphair committed:
update example locally developed app

Co-authored-by: Alexandru Dochioiu <38853913+AlexDochioiu@users.noreply.github.com>

rphair committed:
removed ABNF grammar (now become unpredictable)

Co-authored-by: Alexandru Dochioiu <38853913+AlexDochioiu@users.noreply.github.com>

rphair committed:
update allow/block list behaviour

Co-authored-by: Alexandru Dochioiu <38853913+AlexDochioiu@users.noreply.github.com>

rphair committed:
removed rationale for domain namespacing (which has been removed)

Co-authored-by: Alexandru Dochioiu <38853913+AlexDochioiu@users.noreply.github.com>

rphair committed:
removed rationale for scheme in path (which has been removed)

Co-authored-by: Alexandru Dochioiu <38853913+AlexDochioiu@users.noreply.github.com>

rphair committed:
removed acceptance criterion of dApps needing to publish browse URIs

Co-authored-by: Alexandru Dochioiu <38853913+AlexDochioiu@users.noreply.github.com>

rphair committed:
removed requirement for social testing from Implementation Plan

Co-authored-by: Alexandru Dochioiu <38853913+AlexDochioiu@users.noreply.github.com>

rphair reviewed (commented):
@AlexDochioiu since we've integrated all of your suggestions I think this document can be re-reviewed by wallet representatives (and you can double-check the current state of it).

If no objections (cc @MadOrkestra) we can get ready to merge this... though still hoping for reps from other wallets to comment (by CIP editors' invitation) as requested in https://github.com/cardano-foundation/CIPs/pull/1112#issuecomment-3939590083.

I've bookmarked on my own calendar to review this at the end of next week & mark for `Last Check` at the meeting after that if no such objections have been raised.

Crypto2099 committed:
Bit of cleanup and moving to the correct folder

Crypto2099 commented:
Sorry it has taken me so long to get back to this one. @AlexDochioiu changes looked mostly good and aligned with what has worked well in past demonstrations/VESPR implementation. I pulled changes and did a bit of cleanup for some sections that were affected by the inline/code review edits and this should be ready to proceed. I will update cardano-uri-parser accordingly to reflect this new method.

rphair reviewed (approved):
Thanks @Crypto2099 for coming back on on this.  I do think it's better, all else being equal, to have an ABNF grammar (cc @AlexDochioiu) and I'm happy to have the rest of your updates.

@Ryun1 @perturbing I think it's ready for your commentary in the meanwhile or no-objection at `Last Check` for next meeting (https://hackmd.io/@cip-editors/129) and please @Scitz0 @refi93 @ashisherc @francisluz also indicate if you have any CIP-0158 or general Cardano URI plans that would require adjustment of this CIP candidate.

AlexDochioiu commented:
Everything looks good to me in current state. I have no objections.

Ryun1 reviewed (approved):
discussions have reached consensus
another great contribution to the CIP process and Cardano 💪

MadOrkestra commented:
> We ran out of time at the CIP meeting today to present `Last Check` items: although in the interest of progress and avoiding bureaucratic delays I think we can merge this because of the editor quorum already demonstrated (after a full course of online review).

so close, soooo close :D
